content/courses/advanced-ada/chapters/limited_types.rst

```diff
@@ -6,9 +6,424 @@ Limited Types
+       Put_Line ("--------");
+
+       Put_Line ("Set (B, 7);");
+       Set (B, 7);
+
+       Show (A, "A");
+       Show (B, "B");
+       Show_Compare;
+       Put_Line ("--------");
+
+    end Show_Wrong_Assignment_Equality;
+
+In this code, we declare the :ada:`Simple_Rec` type in the
+:ada:`Nonlimited_Types` package and use it in the
+:ada:`Show_Wrong_Assignment_Equality` procedure. In principle, we're already
+doing many things right here. For example, we're declaring the
+:ada:`Simple_Rec` type private, so that the component :ada:`V` of access
+type is encapsulated. Programmers that declare objects of this type cannot
+simply mess up with the :ada:`V` component. Instead, they have to call the
+:ada:`Init` function and the :ada:`Set` procedure to initialize and change,
+respectively, objects of the :ada:`Simple_Rec` type. That being said, there are
+two problems with this code, which we discuss next.
+
+The first problem we can identify is that the first call to :ada:`Show_Compare`
+shows that :ada:`A` and :ada:`B` are different, although both have the same
+value in the :ada:`V` component (:ada:`A.V.all = 0` and :ada:`B.V.all = 0`)
+|mdash| this was set by the call to the :ada:`Init` function. What's happening
+here is that the :ada:`A = B` expression is comparing the access values
+(:ada:`A.V = B.V`), while we might have been expecting it to compare the actual
+integer values after dereferencing (:ada:`A.V.all = B.V.all`). Therefore, the
+predefined equality function of the :ada:`Simple_Rec` type is useless and
+dangerous for us, as it misleads us to expect something that it doesn't do.
+
+After the assignment of :ada:`A` to :ada:`B` (:ada:`B := A`), the information
+that the application displays seems to be correct |mdash| both :ada:`A.V.all`
+and :ada:`B.V.all` have the same value of two. However, when assigning the
+value seven to :ada:`B` by calling :ada:`Set (B, 7)`, we see that the value of
+:ada:`A.V.all` has also changed. What's happening here is that the previous
+assignment (:ada:`B := A`) has actually assigned access values
+(:ada:`B.V := A.V`), while we might have been expecting it to assign the
+dereferenced values (:ada:`B.V.all := A.V.all`). Therefore, we cannot simply
+directly assign objects of :ada:`Simple_Rec` type, as this operation changes
+the internal structure of the type due to the presence of components of access
+type.
+
+For these reasons, forbidding these operations for the :ada:`Simple_Rec` type
+is the most appropriate software design decision. If we still need assignment
+and equality operators, we can implement custom subprograms for the limited
+type. We'll discuss this topic in the next sections.
+
+In addition to the case when we have components of access types, the
+recommendation to declare a record type limited is also valid when we have
+components of task or protected types. In addition to that, limited types are
+useful when we want to avoid that the same information is copied to multiple
```

Comment:
Probably should be "useful when we want to avoid the situation in which the same information is" or something similar.